Overhaul the events and hook callback system

Murderlon opened the pull request:
Closes #336 
Refs: #309 

- Remove `EVENT_ENDPOINT_CREATED`, `EVENT_FILE_CREATED`, `EVENT_UPLOAD_COMPLETE`, `EVENT_FILE_DELETED`
- Create `POST_CREATE`, `POST_RECEIVE`, `POST_FINISH`, `POST_TERMINATE`.
- Create blocking callbacks `onUploadCreate` and `onUploadFinish`. Error returned will be send back to the client.

Currently the error thrown in the callbacks expects an object `{body: 'some message', status_code: 500}`, which is consistent with the rest of our errors and has a fallback here:

https://github.com/tus/tus-node-server/blob/accb18b2f53d113ded2e176213f353ccc4de8f48/lib/Server.ts#L176-L184

This will be properly documented. But as I'm completely rewriting the readme and splitting up the packages, I'll do that when I get to it.

aduh95 commented on lib/handlers/PatchHandler.ts:
```diff
@@ -30,13 +30,13 @@ export default class PatchHandler extends BaseHandler {
     // with the 404 Not Found or 410 Gone status. The latter one SHOULD
     // be used if the Server is keeping track of expired uploads.
-    const creation = file.creation_date ? new Date(file.creation_date) : new Date()
+    const creation = upload.creation_date ? new Date(upload.creation_date) : new Date()
     const expiration = new Date(creation.getTime() + this.store.getExpiration())
     const now = new Date()
```
Creating `Date` object is quite expensive.
```suggestion
    const now = Date.now()
    const creation = upload.creation_date ? new Date(upload.creation_date).getTime() : now
    const expiration = creation + this.store.getExpiration()
```

Acconut commented:
This PR looks amazing, great work! Only two comments, that can also be addressed in the future. There is no event to obtain the upload progress, comparable to post-receive in tusd. Also, people often want to modify the response in the onUploadCreate and onUploadFinish callbacks (e.g. add custom headers for the client). But the callbacks in this PR only get passed the HTTP request, not the response. 

Both of these can be added in the future, but do you have intentions to address this with the framework laid by this PR?

Murderlon commented:
> But the callbacks in this PR only get passed the HTTP request, not the response.

Changed it pass it the response and also always return the response, modified or not. 

> There is no event to obtain the upload progress, comparable to post-receive in tusd

Added it to the patch handler
